src/server/handlers.ts

```diff
@@ -0,0 +1,1655 @@
+                            }
+
+                            const uri = URI.file(dir).toString();
+                            const context = addContext(
+                                uri,
+                                overrideMotokoVersion,
+                            );
+
+                            try {
+                                context.packages = await getPackageSources(dir);
+                                context.packages.forEach(
+                                    ([name, relativePath]) => {
+                                        const path = resolveVirtualPath(
+                                            uri,
+                                            relativePath,
+                                        );
+                                        console.log(
+                                            'Package:',
+                                            name,
+                                            '->',
+                                            path,
+                                            `(${uri})`,
+                                        );
+                                        context.motoko.usePackage(name, path);
+                                    },
+                                );
+                            } catch (err) {
+                                connection.sendNotification(ERROR_MESSAGE, {
+                                    message: `Error while resolving Motoko packages:`,
+                                    detail: String(err).replace(/^Error: /, ''),
+                                });
+                                context.error = String(err);
+                                console.warn(err);
+                                return;
+                            }
+                        } catch (err: any) {
+                            connection.sendNotification(ERROR_MESSAGE, {
+                                message: `Error while loading Motoko packages:`,
+                                detail: String(err).replace(/^Error: /, ''),
+                            });
+                            console.error(
+                                `Error while reading packages for directory (${dir}): ${err}`,
+                            );
+                            return;
+                        }
+                    }),
+                );
+
+                // Add base library autocompletions
+                // TODO: possibly refactor into `context.ts`
+                Object.entries(baseLibrary.files).forEach(
+                    ([path, { content }]: [string, { content: string }]) => {
+                        notifyWriteUri(`mo:base/${path}`, content);
+                    },
+                );
+
+                loadingPackages = false;
+                notifyWorkspace(); // Update virtual file system
+                notifyDfxChange(); // Reload dfx.json
+                connection.sendNotification('custom/initialized', {});
```

Comment:
I see why this hack is needed, but please add some comments on why it's needed. Someone who is not in the context of the benchmarks could be surprised by this.